server/api.go

```diff
@@ -180,59 +182,141 @@ func (p *Plugin) handleGetAllChannels(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (p *Plugin) handleEndCall(w http.ResponseWriter, r *http.Request, channelID string) {
+	var res httpResponse
+	defer p.httpAudit("handleEndCall", &res, w, r)
+
+	userID := r.Header.Get("Mattermost-User-Id")
+
+	isAdmin := p.API.HasPermissionTo(userID, model.PermissionManageSystem)
+
+	state, err := p.kvGetChannelState(channelID)
+	if err != nil {
+		res.Err = err.Error()
+		res.Code = http.StatusInternalServerError
+		return
+	}
+
+	if state == nil || state.Call == nil {
+		res.Err = "no call ongoing"
+		res.Code = http.StatusBadRequest
+		return
+	}
+
+	if !isAdmin && state.Call.CreatorID != userID {
+		res.Err = "no permissions to end the call"
+		res.Code = http.StatusForbidden
+		return
+	}
+
+	callID := state.Call.ID
+
+	if err := p.kvSetAtomicChannelState(channelID, func(state *channelState) (*channelState, error) {
+		if state == nil || state.Call == nil {
+			return nil, nil
+		}
+
+		if state.Call.ID != callID {
+			return nil, fmt.Errorf("previous call has ended and new one has started")
+		}
+
+		if state.Call.EndAt == 0 {
+			state.Call.EndAt = time.Now().UnixMilli()
+		}
+
+		return state, nil
+	}); err != nil {
+		res.Err = err.Error()
+		res.Code = http.StatusForbidden
+		return
+	}
+
+	p.API.PublishWebSocketEvent(wsEventCallEnd, map[string]interface{}{}, &model.WebsocketBroadcast{ChannelId: channelID, ReliableClusterSend: true})
+
+	go func() {
+		// We wait a few seconds for the call to end cleanly. If this doesn't
+		// happen we force end it.
+		time.Sleep(5 * time.Second)
+
+		state, err := p.kvGetChannelState(channelID)
+		if err != nil {
+			p.LogError(err.Error())
+			return
+		}
+		if state == nil || state.Call == nil || state.Call.ID != callID {
+			return
+		}
+
+		p.LogInfo("call state is still in store, force ending it")
+
+		for connID := range state.Call.Sessions {
+			if err := p.closeRTCSession(userID, connID, channelID, state.NodeID); err != nil {
```

Comment:
~~I kind of wonder why we don't just do this, since the clients can handle it...?~~ Oh I see, the UI modal to notify the users. Got it.

Reply:
We actually could but it's much less clean on the UI as the RTC connection could timeout which would make ending a call feel very "unsynchronized" between clients.